File: templates/default/go/sample/02_control_flow.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {
	// ==========================================
	// 制御構文
	// ==========================================

	fmt.Println("=== Goの制御構文 ===")

	// ------------------------------------------
	// if文
	// ------------------------------------------

	fmt.Println("\n=== if文 ===")

	score := 85

	// 基本的なif文
	if score >= 90 {
		fmt.Println("優秀です！")
	} else if score >= 70 {
		fmt.Println("良い成績です。")
	} else if score >= 50 {
		fmt.Println("合格です。")
	} else {
		fmt.Println("がんばりましょう。")
	}

	// 初期化付きif文
	if grade := calculateGrade(score); grade == "A" {
		fmt.Printf("成績は%sです（素晴らしい！）\n", grade)
	} else {
		fmt.Printf("成績は%sです\n", grade)
	}

	// 複雑な条件
	temperature := 25
	isSunny := true

	if temperature > 20 && isSunny {
		fmt.Println("散歩にいい天気です！")
	} else if temperature > 20 || isSunny {
		fmt.Println("まあまあの天気です。")
	} else {
		fmt.Println("家にいましょう。")
	}

	// ------------------------------------------
	// for文
	// ------------------------------------------

	fmt.Println("\n=== for文 ===")

	// 基本的なfor文
	fmt.Print("カウントアップ: ")
	for i := 0; i < 5; i++ {
		fmt.Printf("%d ", i)
	}
	fmt.Println()

	// while文の代わり
	count := 0
	fmt.Print("while風: ")
	for count < 3 {
		fmt.Printf("%d ", count)
		count++
	}
	fmt.Println()

	// 無限ループ（breakで抜ける）
	counter := 0
	fmt.Print("無限ループ: ")
	for {
		if counter >= 3 {
			break
		}
		fmt.Printf("%d ", counter)
		counter++
	}
	fmt.Println()

	// continue文
	fmt.Print("偶数のみ: ")
	for i := 0; i < 10; i++ {
		if i%2 != 0 {
			continue
		}
		fmt.Printf("%d ", i)
	}
	fmt.Println()

	// ------------------------------------------
	// range文
	// ------------------------------------------

	fmt.Println("\n=== range文 ===")

	// スライスに対するrange
	fruits := []string{"apple", "banana", "cherry", "date"}

	fmt.Println("果物リスト:")
	for index, fruit := range fruits {
		fmt.Printf("%d: %s\n", index, fruit)
	}

	// インデックスのみ使用
	fmt.Print("インデックスのみ: ")
	for index := range fruits {
		fmt.Printf("%d ", index)
	}
	fmt.Println()

	// 値のみ使用
	fmt.Print("値のみ: ")
	for _, fruit := range fruits {
		fmt.Printf("%s ", fruit)
	}
	fmt.Println()

	// マップに対するrange
	ages := map[string]int{
		"Alice": 30,
		"Bob":   25,
		"Carol": 35,
	}

	fmt.Println("年齢リスト:")
	for name, age := range ages {
		fmt.Printf("%s: %d歳\n", name, age)
	}

	// 文字列に対するrange（runeとして処理）
	text := "Hello, 世界"
	fmt.Println("文字列の文字:")
	for i, r := range text {
		fmt.Printf("位置 %d: %c (Unicode: U+%04X)\n", i, r, r)
	}

	// ------------------------------------------
	// switch文
	// ------------------------------------------

	fmt.Println("\n=== switch文 ===")

	day := 3

	// 基本的なswitch文
	switch day {
	case 1:
		fmt.Println("月曜日")
	case 2:
		fmt.Println("火曜日")
	case 3:
		fmt.Println("水曜日")
	case 4:
		fmt.Println("木曜日")
	case 5:
		fmt.Println("金曜日")
	case 6, 7: // 複数の値
		fmt.Println("週末")
	default:
		fmt.Println("無効な日付")
	}

	// 初期化付きswitch
	switch dayType := getDayType(day); dayType {
	case "weekday":
		fmt.Println("平日です")
	case "weekend":
		fmt.Println("週末です")
	default:
		fmt.Println("不明な日付タイプ")
	}

	// 条件付きswitch
	switch {
	case score >= 90:
		fmt.Println("A級")
	case score >= 80:
		fmt.Println("B級")
	case score >= 70:
		fmt.Println("C級")
	default:
		fmt.Println("D級")
	}

	// 型switch
	var value interface{} = 42

	switch v := value.(type) {
	case int:
		fmt.Printf("整数: %d\n", v)
	case string:
		fmt.Printf("文字列: %s\n", v)
	case bool:
		fmt.Printf("論理値: %t\n", v)
	default:
		fmt.Printf("不明な型: %T\n", v)
	}

	// fallthroughの使用例
	grade := "B"
	fmt.Print("成績評価: ")
	switch grade {
	case "A":
		fmt.Print("優秀 ")
		fallthrough
	case "B":
		fmt.Print("良好 ")
		fallthrough
	case "C":
		fmt.Print("普通 ")
		fallthrough
	default:
		fmt.Print("以上")
	}
	fmt.Println()

	// ------------------------------------------
	// ラベル付きbreak/continue
	// ------------------------------------------

	fmt.Println("\n=== ラベル付きbreak/continue ===")

	// ネストしたループからの脱出
	fmt.Println("ラベル付きbreak:")
outer:
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if i == 1 && j == 1 {
				fmt.Println("外側のループを抜けます")
				break outer
			}
			fmt.Printf("(%d, %d) ", i, j)
		}
		fmt.Println()
	}

	// ラベル付きcontinue
	fmt.Println("ラベル付きcontinue:")
outerContinue:
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if j == 1 {
				fmt.Print("外側の次の反復へ ")
				continue outerContinue
			}
			fmt.Printf("(%d, %d) ", i, j)
		}
		fmt.Println()
	}

	// ------------------------------------------
	// defer文
	// ------------------------------------------

	fmt.Println("\n=== defer文 ===")

	// 基本的なdefer
	fmt.Println("処理開始")
	defer fmt.Println("処理終了（defer）")
	defer fmt.Println("クリーンアップ（defer）")

	fmt.Println("メイン処理")

	// deferは後入れ先出し（LIFO）で実行される
	for i := 0; i < 3; i++ {
		defer fmt.Printf("defer %d\n", i)
	}

	// deferで関数呼び出し
	deferExample()

	// ------------------------------------------
	// select文（基本）
	// ------------------------------------------

	fmt.Println("\n=== select文（基本） ===")

	// チャネルの基本例（詳細は後の章で）
	ch1 := make(chan string, 1)
	ch2 := make(chan string, 1)

	ch1 <- "チャネル1からのメッセージ"

	select {
	case msg1 := <-ch1:
		fmt.Printf("受信: %s\n", msg1)
	case msg2 := <-ch2:
		fmt.Printf("受信: %s\n", msg2)
	default:
		fmt.Println("どのチャネルも準備できていません")
	}

	// タイムアウト付きselect
	timeout := time.After(100 * time.Millisecond)
	select {
	case <-timeout:
		fmt.Println("タイムアウトしました")
	default:
		fmt.Println("すぐに実行されました")
	}

	// ------------------------------------------
	// 実践例：簡単なゲーム
	// ------------------------------------------

	fmt.Println("\n=== 実践例：数当てゲーム ===")

	rand.Seed(time.Now().UnixNano())
	target := rand.Intn(10) + 1
	maxAttempts := 3

	fmt.Printf("1から10の数を当ててください（%d回まで）\n", maxAttempts)

	for attempt := 1; attempt <= maxAttempts; attempt++ {
		// 実際のゲームでは標準入力から読み取りますが、
		// ここでは乱数でシミュレート
		guess := rand.Intn(10) + 1
		fmt.Printf("試行 %d: %d\n", attempt, guess)

		switch {
		case guess == target:
			fmt.Println("正解です！おめでとうございます！")
			return
		case guess < target:
			fmt.Println("もっと大きい数です")
		case guess > target:
			fmt.Println("もっと小さい数です")
		}

		if attempt == maxAttempts {
			fmt.Printf("残念！正解は %d でした\n", target)
		}
	}
}

// ------------------------------------------
// ヘルパー関数
// ------------------------------------------

func calculateGrade(score int) string {
	switch {
	case score >= 90:
		return "A"
	case score >= 80:
		return "B"
	case score >= 70:
		return "C"
	case score >= 60:
		return "D"
	default:
		return "F"
	}
}

func getDayType(day int) string {
	switch day {
	case 1, 2, 3, 4, 5:
		return "weekday"
	case 6, 7:
		return "weekend"
	default:
		return "invalid"
	}
}

func deferExample() {
	fmt.Println("関数開始")
	defer fmt.Println("関数終了（defer）")

	for i := 0; i < 3; i++ {
		defer func(n int) {
			fmt.Printf("ループdefer: %d\n", n)
		}(i) // クロージャで値をキャプチャ
	}

	fmt.Println("関数の主要処理")
}

File: learning/zig/02_control_flow.zig
const std = @import("std");

pub fn main() !void {
    // ==========================================
    // 制御構文
    // ==========================================

    std.debug.print("=== Zigの制御構文 ===\n", .{});

    // ------------------------------------------
    // if文
    // ------------------------------------------

    std.debug.print("\n=== if文 ===\n", .{});

    const score = 85;

    // 基本的なif文
    if (score >= 90) {
        std.debug.print("優秀です！\n", .{});
    } else if (score >= 70) {
        std.debug.print("良い成績です。\n", .{});
    } else if (score >= 50) {
        std.debug.print("合格です。\n", .{});
    } else {
        std.debug.print("がんばりましょう。\n", .{});
    }

    // if式（値を返す）
    const grade = if (score >= 90) "A" else if (score >= 80) "B" else if (score >= 70) "C" else "D";
    std.debug.print("成績: {s}\n", .{grade});

    // Optional型とif文の組み合わせ
    const maybe_value: ?i32 = 42;
    if (maybe_value) |value| {
        std.debug.print("値が存在します: {d}\n", .{value});
    } else {
        std.debug.print("値がありません\n", .{});
    }

    // ------------------------------------------
    // while文
    // ------------------------------------------

    std.debug.print("\n=== while文 ===\n", .{});

    // 基本的なwhile文
    var i: u32 = 0;
    std.debug.print("カウントアップ: ", .{});
    while (i < 5) {
        std.debug.print("{d} ", .{i});
        i += 1;
    }
    std.debug.print("\n", .{});

    // while文でOptional型を扱う
    const numbers = [_]i32{ 1, 2, 3, 4, 5 };
    var index: usize = 0;

    std.debug.print("配列の要素: ", .{});
    while (index < numbers.len) : (index += 1) {
        std.debug.print("{d} ", .{numbers[index]});
    }
    std.debug.print("\n", .{});

    // continueラベル付きwhile
    var counter: u32 = 0;
    std.debug.print("偶数のみ: ", .{});
    while (counter < 10) : (counter += 1) {
        if (counter % 2 != 0) continue;
        std.debug.print("{d} ", .{counter});
    }
    std.debug.print("\n", .{});

    // ------------------------------------------
    // for文
    // ------------------------------------------

    std.debug.print("\n=== for文 ===\n", .{});

    // 配列に対するfor文
    const fruits = [_][]const u8{ "apple", "banana", "cherry", "date" };

    std.debug.print("果物リスト:\n", .{});
    for (fruits) |fruit| {
        std.debug.print("- {s}\n", .{fruit});
    }

    // インデックス付きfor文
    std.debug.print("インデックス付き:\n", .{});
    for (fruits, 0..) |fruit, idx| {
        std.debug.print("{d}: {s}\n", .{ idx, fruit });
    }

    // 範囲for文（スライス）
    const slice = fruits[1..3];
    std.debug.print("スライス要素: ", .{});
    for (slice) |item| {
        std.debug.print("{s} ", .{item});
    }
    std.debug.print("\n", .{});

    // for文による配列の変更
    var mutable_numbers = [_]i32{ 1, 2, 3, 4, 5 };
    std.debug.print("変更前: ", .{});
    for (mutable_numbers) |num| {
        std.debug.print("{d} ", .{num});
    }
    std.debug.print("\n", .{});

    // ポインタを使って配列の要素を変更
    for (&mutable_numbers) |*num| {
        num.* *= 2;
    }

    std.debug.print("変更後: ", .{});
    for (mutable_numbers) |num| {
        std.debug.print("{d} ", .{num});
    }
    std.debug.print("\n", .{});

    // ------------------------------------------
    // switch文
    // ------------------------------------------

    std.debug.print("\n=== switch文 ===\n", .{});

    const day = 3;

    // 基本的なswitch文
    switch (day) {
        1 => std.debug.print("月曜日\n", .{}),
        2 => std.debug.print("火曜日\n", .{}),
        3 => std.debug.print("水曜日\n", .{}),
        4 => std.debug.print("木曜日\n", .{}),
        5 => std.debug.print("金曜日\n", .{}),
        6, 7 => std.debug.print("週末\n", .{}), // 複数の値をまとめて処理
        else => std.debug.print("無効な日付\n", .{}),
    }

    // switch式（値を返す）
    const day_type = switch (day) {
        1...5 => "平日",
        6, 7 => "週末",
        else => "無効",
    };
    std.debug.print("日付タイプ: {s}\n", .{day_type});

    // 文字に対するswitch
    const letter = 'A';
    switch (letter) {
        'A', 'E', 'I', 'O', 'U' => std.debug.print("'{c}' は母音です\n", .{letter}),
        'a', 'e', 'i', 'o', 'u' => std.debug.print("'{c}' は母音です（小文字）\n", .{letter}),
        else => std.debug.print("'{c}' は子音です\n", .{letter}),
    }

    // ------------------------------------------
    // ネストしたループとラベル
    // ------------------------------------------

    std.debug.print("\n=== ネストしたループとラベル ===\n", .{});

    // ラベル付きループ
    outer: for (0..3) |i_outer| {
        for (0..3) |j_inner| {
            if (i_outer == 1 and j_inner == 1) {
                std.debug.print("外側のループを終了\n", .{});
                break :outer;
            }
            std.debug.print("({d}, {d}) ", .{ i_outer, j_inner });
        }
    }
    std.debug.print("\n", .{});

    // continueラベル
    std.debug.print("continueラベルの例:\n", .{});
    outer_continue: for (0..3) |i_outer| {
        for (0..3) |j_inner| {
            if (j_inner == 1) {
                std.debug.print("外側のループの次の反復へ\n", .{});
                continue :outer_continue;
            }
            std.debug.print("({d}, {d}) ", .{ i_outer, j_inner });
        }
        std.debug.print("\n", .{});
    }

    // ------------------------------------------
    // defer文
    // ------------------------------------------

    std.debug.print("\n=== defer文 ===\n", .{});

    {
        std.debug.print("処理開始\n", .{});
        defer std.debug.print("処理終了（defer）\n", .{});
        defer std.debug.print("cleanup処理（defer）\n", .{});

        std.debug.print("メイン処理\n", .{});
        // スコープを抜ける際に、defer文が逆順で実行される
    }

    // deferは関数レベルでも使用可能
    defer_example();
}

fn defer_example() void {
    std.debug.print("関数開始\n", .{});
    defer std.debug.print("関数終了（defer）\n", .{});

    var i: u32 = 0;
    while (i < 3) {
        defer std.debug.print("ループ終了 {d}（defer）\n", .{i});
        std.debug.print("ループ {d}\n", .{i});
        i += 1;
    }
}
